qiskit/synthesis/multi_controlled/mcx_synthesis.py

```diff
@@ -230,3 +233,113 @@ def synth_mcx_1_clean_b95(num_ctrl_qubits: int):
     )
 
     return qc
+
+
+def synth_mcx_gray_code(num_ctrl_qubits: int) -> QuantumCircuit:
+    r"""
+    Synthesize a multi-controlled X gate with :math:`k` controls using the Gray code.
+
+    Produces a quantum circuit with :math:`num_ctrl_qubits + 1` qubits.
+
+    Args:
+        num_ctrl_qubits: The number of control qubits.
+
+    Returns:
+        The synthesized quantum circuit.
+    """
+    num_qubits = num_ctrl_qubits + 1
+    q = QuantumRegister(num_qubits, name="q")
+    qc = QuantumCircuit(q, name="mcx_gray")
+    qc._append(HGate(), [q[-1]], [])
+    qc._append(MCU1Gate(np.pi, num_ctrl_qubits=num_ctrl_qubits), q[:], [])
+    qc._append(HGate(), [q[-1]], [])
+    return qc
+
+
+def synth_mcx_mcphase(num_ctrl_qubits: int) -> QuantumCircuit:
+    r"""
+    Synthesize a multi-controlled X gate with :math:`k` controls based on
+    the implementation for MCPhaseGate.
+
+    Produces a quantum circuit with :math:`num_ctrl_qubits + 1` qubits.
+
+    Args:
+        num_ctrl_qubits: The number of control qubits.
+
+    Returns:
+        The synthesized quantum circuit.
+    """
+    if num_ctrl_qubits == 3:
+        return synth_c3x()
+
+    if num_ctrl_qubits == 4:
+        return synth_c4x()
+
+    num_qubits = num_ctrl_qubits + 1
+    q = QuantumRegister(num_qubits, name="q")
+    qc = QuantumCircuit(q)
+    q_controls = list(range(num_ctrl_qubits))
+    q_target = num_ctrl_qubits
+    qc.h(q_target)
+    qc.mcp(np.pi, q_controls, q_target)
+    qc.h(q_target)
+    return qc
+
+
+def synth_c3x() -> QuantumCircuit:
```

Comment:
why should we copy the synthesis code here and not use C3XGate and C4XGate directly?

Reply:
I think the problem is that `C3XGate` and `C4XGate` are called `"mcx"` which would lead to an infinite recursion in the HLS plugin. Also I'm not sure what the benefit of having these gates is, it might be best to deprecate them as well in favor of just using `MCXGate` -- what do you think?

If we want to keep them, an alternative solution to the infinite-recursion could be to instance-check on `C3XGate/C4XGate` and return `None`, which will then trigger the unroller.

Reply:
We discussed this offline with @Cryoris and thought that calling `synth_c3x()` and `synth_c4x()` instead of using `C3XGate` and `C4XGate` directly is a slightly better approach. A more minor point is that this decreases the level of "nestedness" and should be slightly faster. A more important problem (as I have already complained in the PR summary) is that C3XGate and C4XGate have name `"mcx"`, so they are caught by `HighLevelSynthesis` plugins for MCX-gates (which I think they should). (Note: fortunately CX and CCX gates are _not_ called `"mcx"`). In order to avoid recursion, we need to ensure that a synthesis plugin for MCX gates simplifies the logic, i.e. it may replace a 10-controlled MCX-gate by some smaller MCX-gates, but it should not replace a `C3XGate` by a `C3XGate`. This is exactly what was happening before. We could have changed this both on the plugin level or the synthesis function level, but thought that the second option is slightly cleaner. I can also do it the other way around, if you think that's better.